runtime/interp.c

```diff
@@ -1372,7 +1376,9 @@ do_resume: {
       if (parent == NULL) {
         accu = caml_continuation_use(cont);
         resume_fn = raise_unhandled_effect;
+        Setup_for_gc;
         resume_arg = caml_make_unhandled_effect_exn(eff);
+        Restore_after_gc;
```

Comment:
I think there's another problem here: `resume_fn` is not a root, so if a GC occurs just after, the closure `raise_unhandled_effect` can be moved and `resume_fn` is invalidated.

Here is a suggestion that addresses both this problem and the problem with `accu`:

```suggestion
        Setup_for_c_call;
        resume_arg = caml_make_unhandled_effect_exn(eff);
        accu = caml_continuation_use(cont);
        Restore_after_c_call;
        resume_fn = raise_unhandled_effect;

```
As far as I can see, `caml_continuation_use` doesn't allocate in the heap, at least when it returns normally.  However, it can raise an exception, which is why I moved it inside the Setup/Restore section.  (For an esception raised by C to be trapped by the bytecode interpreter, the `Idomain_state->current_stack->sp` must contain the correct `sp` value; the `Setup_for_*` macros take care of that.)

Reply:
That gets those proverbial two birds with one stone, thanks! 🙂

It's clearly better to have the code such that we know `resume_fn` can't be invalidated. Just for my understanding, I think it's the case (without a compactor) that it couldn't have moved before, though - `raise_unhandled_effect` is a global root and the value is set-up when the interpreter is initialised and before the first minor collection at startup. It would therefore have always been in a fixed place in the major heap when the interpreter actually started the interpreting the program?

Reply:
You're probably right, because there's no compactor and a minor GC moved the closure to the major heap earlier.  But this is a complicated reasoning that I don't want to make, ever.  Any heap-allocated value can move during any allocation, that's a simple rule.